mm/alignment.py

```diff
@@ -789,3 +792,176 @@ def plot(self):
+            segment_id_to_alignment[seq_tag] = " ".join(segment_id_to_alignment[seq_tag])
+
+        # Get the corpus information: segment_id -> text.
+        c = corpus.Corpus()
+        c.load(self.corpus_file.get_path())
+        segment_id_to_text = {segment.fullname(): segment.orth for segment in c.segments()}
+
+        with util.uopen(self.out_text_alignment_pairs.get_path(), "wt") as f:
+            f.write(f"segment_id{self.csv_separator}segment_text{self.csv_separator}segment_alignment\n")
+            for segment_id in set(segment_id_to_alignment.keys()).intersection(set(segment_id_to_text.keys())):
+                f.write(
+                    f"{segment_id}{self.csv_separator}"
+                    f"{segment_id_to_text[segment_id]}{self.csv_separator}"
+                    f"{segment_id_to_alignment[segment_id]}{self.csv_separator}\n"
+                )
+
+
+class PlotViterbiAlignmentJob(Job):
+    """
+    Plots the alignments of each segment in the specified alignment files.
+
+    Original job author: dmann.
+    """
+
+    def __init__(self, alignment_files: Iterable[tk.Path], allophone_file: tk.Path):
+        self.alignment_files = alignment_files
+        self.allophone_file = allophone_file
+
+        self.out_plot_dir = self.output_path("plots", directory=True)
+        self.out_empty_alignment_seq_tags = self.output_path("empty_alignment_seq_tags.txt")
+
+        self.rqmt = {"cpu": 1, "mem": 2.0, "time": 1.0}
+
+    def tasks(self):
+        yield Task("run", resume="run", rqmt=self.rqmt, args=range(1, len(self.alignment_files) + 1))
+
+    def extract_phoneme_sequence(self, alignment: np.array) -> Tuple[np.array, np.array]:
+        """
+        :param alignment: Monophone alignment, for instance: `np.array(["a", "a", "b", ...])`.
+        :return: Monophone sequence (ordered as given),
+            as well as the indices corresponding to the monophone sequence from the Viterbi alignment.
+        """
+        boundaries = np.concatenate(
+            [
+                np.where(alignment[:-1] != alignment[1:])[0],
+                [len(alignment) - 1],  # manually add boundary of last allophone
+            ]
+        )
+
+        lengths = boundaries - np.concatenate([[-1], boundaries[:-1]])
+        phonemes = alignment[boundaries]
+        monotonic_idx_alignment = np.repeat(np.arange(len(phonemes)), lengths)
+        return phonemes, monotonic_idx_alignment
+
+    def make_viterbi_matrix(self, label_idx_seq: np.array) -> np.array:
+        """
+        :return: Matrix corresponding to the Viterbi alignment.
+        """
+        num_alignments = len(label_idx_seq)
+        max_timestamp = max(label_idx_seq) + 1
+        viterbi_matrix = np.zeros((max_timestamp, num_alignments), dtype=np.float32)
+        for t, idx in enumerate(label_idx_seq):
+            viterbi_matrix[idx, t] = 1.0
+        return viterbi_matrix
+
+    def plot(self, viterbi_matrix: np.array, allophone_sequence: List[str]):
```

Comment:
The typing indicates that the function does not return anything, but this is not the case.